options.go

```diff
@@ -123,6 +123,15 @@ func PostBuild(fn func(*Kong) error) Option {
 	})
 }
 
+func BeforeApplyHook(fn hook) Option {
```

Comment:
I was literally just wanting this functionality yesterday, so +1 from me.

I'd prefer this to be named WithBeforeApply, and fn will have to be any to correctly match any function.

Reply:
Done!

What do you think of removing https://github.com/alecthomas/kong/blob/master/hooks.go and naming these functional options `BeforeReset`, `BeforeResolve`, etc?